File: 905.sort-array-by-parity.java
/*
 * @lc app=leetcode id=905 lang=java
 *
 * [905] Sort Array By Parity
 */

// @lc code=start
class Solution {
    public int[] sortArrayByParity(int[] nums) {
        if (nums.length == 1) {
            return nums;
        }
        int left = 0, right = nums.length - 1;
        while (left < right) {
            while (left < right && nums[left] % 2 == 0) {
                left += 1;
            }
            while (left < right && nums[right] % 2 == 1) {
                right -= 1;
            }
            int tmp = nums[left];
            nums[left] = nums[right];
            nums[right] = tmp;
            left += 1;
            right -= 1;
        }
        return nums;
    }
}
// @lc code=end

File: 905.sort-array-by-parity.cpp
/*
 * @lc app=leetcode id=905 lang=cpp
 *
 * [905] Sort Array By Parity
 */

// @lc code=start
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        for (auto l = nums.begin(), r = nums.end() - 1; l <= r; ) {
            if (*l % 2 == 0) ++l;
            else if (*r % 2 == 1) --r;
            else {
                swap(*l, *r); ++l; --r;
            }
        }
        return nums;
    }
};
// @lc code=end
